[ADD] estate: created estate module

danal-odoo opened the pull request:
1. estate module
    Create new module estate module for manage estate property with feature like create new property manage buy and sell with usererror and database constraint to reduce miss matched data to be enter in DB.
2. Awesome OWL
    Create Card, Counter and Todo component.
3. Awesome Dashboard
   add customers and leads button in layout and added dynamic card layout to render items with chartcard  and numbercard. dashboard data will be refresh every 10 mins and added lazyloading to Dashboard.

danal-odoo committed:
[ADD] estate: created estate application and important files.

added estate module.
created __init__.py and __manifest__.py files.

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1022.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1022)

danal-odoo committed:
[IMP] estate: created models for estate application.

created estate property model.
added all required fields in model.

danal-odoo committed:
[IMP] estate: implimented security for model access.

added security CSV file in security folder.
added estate_property model access to user in csv file.

danal-odoo committed:
[IMP] estate: created views, menuitem and modified manifest file.

added action view in estate property.
created menuitems and linked action to that menuitems.
updated manifest file to link all created file.

danal-odoo committed:
[IMP] estate: implemented form and model fields, code formation.

implemented form in views and added model fields in estate_property model.
modified views and menuitem names.

danal-odoo committed:
[IMP] estate: added form and search with Filter and Group By functionality.

added estate property form and list view.
added search with filter and group_by functionality.

danal-odoo committed:
[IMP] estate: added offer, tag and type models and created views.

added estate_property_offer model and created required fields.
added estate_property_tag model and created required fields.
added estate_property_type model and created required fields.
added models acces to user.

danal-odoo committed:
[IMP] estate: add compute, inverse and on chnage in estate offer model.

added @api.depends() decorator to calculate total area and best prrice.
added @api.onchange() decorator to auto fill garden_area and orientation on garden change.

bit-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    "application": True,
+    "license": "LGPL-3",
+    "description": """
+    Description text
```
Should be a proper description.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python3
```
It is not necessary to add.

bit-odoo commented on estate/security/ir.model.access.csv:
```diff
@@ -0,0 +1,5 @@
+estate.access_estate_property,"access_estate_property",estate.model_estate_property,base.group_user,1,1,1,1
+estate.access_estate_property_type,"access_estate_property_type",estate.model_estate_property_type,base.group_user,1,1,1,1
+estate.access_estate_property_tag,"access_estate_property_tag",estate.model_estate_property_tag,base.group_user,1,1,1,1
+estate.access_estate_property_offer,"access_estate_property_offer",estate.model_estate_property_offer,base.group_user,1,1,1,1
```
There should be one empty line at the end of the file.

bit-odoo commented on estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,41 @@
+        id="estate_menu_root" 
+        name="Estate Property"
+    /> 
+
```
unnecessary space.

bit-odoo commented on estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,41 @@
+        parent="estate_menu_root"
+        sequence="1"
+    />
+
```
same.

bit-odoo commented on estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,41 @@
+        parent="estate_menu_configuration"
+        action="action_estate_property_tag"
+    />
+</odoo> 
```
There should be one empty line at the end of the file.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,161 @@
+                </search>
+            </field>
+    </record>
+</odoo>
```
There should be one empty line at the end of the file.

bit-odoo reviewed (commented):
Hello!
Nice work.
i have added some comments please have a look.

Can you please adapt your commit message and you need to add PR title and description.

you can take reference from here for commit message and title :- https://www.odoo.com/documentation/19.0/contributing/development/git_guidelines.html#commit-message-full-description

danal-odoo committed:
[IMP] estate: added restrictions and action in property offer.

added action method on sold and cancel button.
added UserError() to prevent sold property to cancel.
added UserError() to prevent cancelled property to be sold.

bit-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,46 @@
+    def action_refuse(self):
+        for record in self:
+            record.status = "refused"
+            record.property_id.selling_price = 0.00
```
Why do we change the selling price when refusing offer?

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,94 @@
+    _name = "estate.property"
+    _description = "Estate Property"
+
+    name = fields.Char(required=True, default="Unknown")
```
Why default `Unknown` ?

bit-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    "category": "Category",
+    "application": True,
+    "license": "LGPL-3",
+    "description": """
```
I feel that description should be after the name. It's only my suggestion.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,94 @@
+    garage = fields.Boolean()
+    garden = fields.Boolean()
+    garden_area = fields.Integer()
+    last_seen = fields.Datetime("Last Seen", default=fields.Datetime.now)
```
What is the use of this field?

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,94 @@
+    tag_ids = fields.Many2many("estate.property.tag", string="Property Tags")
+    offer_ids = fields.One2many("estate.property.offer", "property_id", string="Offer")
+    total_area = fields.Integer(compute="_compute_area")
+    best_price = fields.Integer(compute="_compute_highest")
```
```suggestion
    best_price = fields.Integer(compute="_compute_best_price")
```
It is a good habit to compute the method name related to the field name. So others can easily identify.

danal-odoo committed:
[IMP] estate: added sql constraints and python constraints, fixed offer view

[FIX] estate: code formatting and rename compute method

Improved code readability by applying consistent formatting.
Renamed compute method for best_price to follow naming conventions. changed discription position after name in manifest file.

bit-odoo reviewed (commented):
Hello,
I have added some comments.

Can you please adapt your commit messages as I informed you in the first review?

danal-odoo committed:
[IMP] estate: added stat button, inline views, list order

added stat button in property type to show quick offer list.
  added inline views in property types to get related fields from other model.
  added order method in models, views, manual.
  improved some logic in percentage calculation.
  code formation.

danal-odoo committed:
[IMP] estate: added inheritance and on create functions.

added python inheritance, model inheritance, view inheritance.
added create function to change state from 'new' to 'offer_received'.
separated all property action to it's own views file.

danal-odoo committed:
[ADD] estate: added Estate Account Module

created Estate Account module and necessary files.
added invoicing functionality by using inherit method.
added UserError on sold button if no offer are accepted.

danal-odoo committed:
[IMP] estate: added kanban view and optimized offer state button

added kanban view in property default group by types.
fixed state button to show offers in number with proper format.
added option attr to all numerical field to prevent user error.
added group by state in search.

danal-odoo committed:
[IMP] estate: code formation as per odoo guidline

danal-odoo committed:
[IMP] awesome_owl: added counter ui in playground, added card,counter component.

added counter button and component in  playground xml.
created counter and card component with it's xml and js files.
added property field in sales order form using inheritance.

danal-odoo committed:
[IMP] awesome_owl: added markup(),props validation, sum counter

added markup() to render html element from component.
added props validation in content in Card component.
added calculateSum function to calculate sum of counter.

bit-odoo commented:
Hello @danal-odoo 
I already informed you twice, but you haven't changed your commit message and PR title.

Can you please explain it to me why you push a separate commit for this type of changes?
Like this - 1a3bc6501523c5d67eb2d1df86572fe77b7b3070, cc1dabdb8548f3cde7eb718b0285ae0d12d9881d

danal-odoo committed:
[IMP] awesome_owl: added todo component.

  added todo component and created todolist and todoitem.
  created function to add todo item by user.
  fixed layout to be look more organize.

danal-odoo committed:
[IMP] awesome_owl: added todo component and made card component dynamic.

created todo component and in that component created todoitem and todolist.
added todoitem in todolist to get all created todoitem with todolist using props.
added todolist in playground.js to render all todolist.
made card component dynamic by using slots prop by seting type to object.
added card toogle button to show card content or not.
made offer action button invisible if one offer is accepted.

danal-odoo committed:
[IMP] awesome_dashboard: added layout component in dashboard.

added layout component in dashboard.js.
created dashboard.scss to use className in xml.
added display prop in layout.

danal-odoo committed:
[IMP] awesome_dashboard: added customers and leads button.

added customers and leads button in controlpanel layout.
linked onclick in button  function to call view action for customers and leads.

danal-odoo committed:
[IMP] awesome_dashboard: added DashboardItem component.

  added DashboardItem component.
  added slot props in DashboardItem and created a dynamic card in xml.
  added rpc call in onWillStart().
  render all return values in it's separate card in dashboard.xml.
  added state as 'badge' widget in kanban view in estate module.
  refined garden onchange logic.

bit-odoo commented on awesome_dashboard/static/src/dashboard.xml:
```diff
@@ -1,8 +1,47 @@
 <?xml version="1.0" encoding="UTF-8" ?>
 <templates xml:space="preserve">
-
```
unneccary diff.

bit-odoo commented on awesome_dashboard/static/src/dashboard.xml:
```diff
@@ -1,8 +1,47 @@
+            </div>
+        </Layout>
     </t>
-
```
unneccary diff.

bit-odoo commented on awesome_owl/static/src/counter/counter.js:
```diff
@@ -0,0 +1,31 @@
+
+    increment() {
+        this.state.value++;
+        if (this.props.onChange){
```
```suggestion
        if (this.props.onChange) {
```

bit-odoo commented on awesome_owl/static/src/playground.js:
```diff
@@ -1,5 +1,30 @@
+    }
+
+    calculateSum(newValue) {
+        if (newValue > 0){
```
```suggestion
        if (newValue > 0) {
```

bit-odoo commented on awesome_owl/static/src/playground.js:
```diff
@@ -1,5 +1,30 @@
+        if (newValue > 0){
+            this.state.sum++;
+        }   
+        if (newValue < 0){
```
```suggestion
        if (newValue < 0) {
```

bit-odoo commented on awesome_owl/static/src/playground.js:
```diff
@@ -1,5 +1,30 @@
+        }
+    }
+
+    reset(){
```
```suggestion
    reset() {
```

bit-odoo commented on awesome_owl/static/src/playground.xml:
```diff
@@ -1,10 +1,41 @@
 <?xml version="1.0" encoding="UTF-8" ?>
 <templates xml:space="preserve">
-
```
unneccary diff.

bit-odoo commented on awesome_owl/static/src/playground.xml:
```diff
@@ -1,10 +1,41 @@
+            </Card>
         </div>
     </t>
-
```
unneccary diff.

bit-odoo commented on awesome_owl/static/src/todo/todoitem.js:
```diff
@@ -0,0 +1,19 @@
+        removeTodo: Function
+    };
+
+    toggleIsCompleted(){
```
```suggestion
    toggleIsCompleted() {
```

bit-odoo commented on awesome_owl/static/src/todo/todolist.js:
```diff
@@ -0,0 +1,39 @@
+        todo.isCompleted = !todo.isCompleted;
+    }
+
+    removeTodo(id){
```
```suggestion
    removeTodo(id) {
```

bit-odoo commented on awesome_owl/static/src/todo/todolist.js:
```diff
@@ -0,0 +1,39 @@
+
+    removeTodo(id){
+        const index = this.todos.findIndex((elem) => elem.id === id);
+        if (index >= 0){
```
```suggestion
        if (index >= 0) {
```

bit-odoo reviewed (commented):
Hello,
I have added some suggestions.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,123 @@
+    )
+    _check_selling_price = models.Constraint(
+        "CHECK(selling_price >= 0)", "A property selling price must be positive."
+    )
```
It is good that we can declare it after the fields.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,123 @@
+    def _onchange_garden(self):
+        self.garden_area = 10
+        self.garden_orientation = "north"
+        if not self.garden:
```
Why multiple times change the value if the garden is not true?. It is unnecessarily writing a value.

Can you please explain it to me?

danal-odoo replied:
so i used inverse condition to avoid use of else in this function but now i understand my mistake if i don't use else then it will first store garden_area = 10 and then orientation = "north" every time this onchange function is called so i have to use if else condition to avoid extra db write.

bit-odoo replied:
Why do you want to avoid the else condition?. Could you please explain it?

danal-odoo replied:
to reduce check condition and make code faster

bit-odoo replied:
Is `else` reduce performance? I don't think it is right.
Currently, your code causes performance issues.

Can you please elaborate properly? So I can understand your thought.

danal-odoo replied:
yes sir current code is causing performance issue cause it is writing 2 fields without any reason those extra write can give huge impact on busy server and i totally forgot about that performance issue in db i focused on making current function more optimized but it was not space efficient and in db  it is very important to use every ORM carefully to reduce impact on dbservers i have revert back to my original optimized logic to fix performance issue.

bit-odoo replied:
Can you please explain to me what you wrote here https://github.com/odoo/tutorials/pull/1022#discussion_r2559661189
`make code faster.`

How `else` cause the performance issue?

danal-odoo replied:
like it will not check if condition first then it is false and then run else instead it will directly assign garden_area = 10 and orientation to "north" cause at first garden is by default false so second condition will not run but will cause issue when garden change from True to False it will rewrite those true condition again so will query 4 time after change of garden so yes it is not faster.

danal-odoo replied:
> Can you please explain to me what you wrote here [#1022 (comment)](https://github.com/odoo/tutorials/pull/1022#discussion_r2559661189) `make code faster.`
> 
> How `else` cause the performance issue?

no sir, else does not reduce performance and it was  my mistake to remove else. infect in this case it increase performance by reducing db query.

danal-odoo committed:
[IMP] awesome_dashboard: added Chart and made dashboard lazyload.

  created service to load rpc call and load rpc data.
  added interval for rpc call to refresh data every 10 minuets.
  added Chart to show t-shirt size in chart view.
  load dashboard as lazy-loading by register that Component to lazy-components.

danal-odoo committed:
[IMP] awesome_dashboard: added popup to select item to show/hide.

added a settings icon near Dashboard in Layout component.
added popup form to select which item to show/hide in Dashboard.
added function which will store those settings in Browser's local storage.

danal-odoo committed:
[IMP] estate: fixed state button layout.

added onchange on offer_ids if  len(offer_ids) is zero then will change
state to 'new'.
fixed state button layout was in header first shifted to context menu.
added sold button disabled when there is no offer accepted and
if offer is accepted then  will show in highlight color.

danal-odoo committed:
[IMP] estate: added bulk offer wizard

before, there was no option to bulk offer to any property.
which is so inconvenient for a salesperson to offer on a property.

this commit adds that feature to bulk offer to property which
`expected_price` is 5 Lac or lower.
and added webcontroller to display all properties in website.
